### DISCUSSION TITLE
Tag helpers

### DISCUSSION THREAD

**Discussion by dharmatech** [giraffe-fsharp/Giraffe#457] [Q&A]
In a [previous post](https://github.com/giraffe-fsharp/Giraffe/discussions/456), I mentioned MvcMovieGiraffe.

Here's a usage of the [input tag helper](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/working-with-forms?view=aspnetcore-5.0#the-input-tag-helper) from the C# version of that project:

    <input asp-for="Title" class="form-control" />

It references the following member of the model:

```
[StringLength(60, MinimumLength = 3), Required]
public string Title { get; set; }
```

If you look at source of the resulting page, that tag helper expands into:

```
<input 
    class="form-control" 
    type="text" 
    data-val="true" 
    data-val-length="The field Title must be a string with a minimum length of 3 and a maximum length of 60." 
    data-val-length-max="60" 
    data-val-length-min="3" 
    data-val-required="The Title field is required." 
    id="Title" 
    maxlength="60" 
    name="Title" 
    value="" />
```

Is there something like tag helpers for Giraffe? Either built-in or available as a library?

I messed around with putting together something like a tag helper ([source](https://gist.github.com/dharmatech/9c3d30bc77d6346b2ec3a5f0f67880b1)). Example call:

    input_helper typedefof<Movie> "Title" "" "Title" [ _class "form-control" ]

That expands into:

```
<input class="form-control" type="text" data-val="true" data-val-length-max="60" id="Title" name="Title" value="">
```

Thanks for any suggestions!

**Comment by dustinmoris** [comment]
Giraffe doesn't have something like tag helpers at the moment and I don't know of any project which adds that functionality to Giraffe, but it's certainly something which is possible. It would have to use reflection, but it shouldn't be too difficult for a function to accept a a model and the name of property and then do the magic. The only issue I think we won't get away (even with F# trickery) is the fact that one has to pass the name of the property as a "magic" string into such a function, but given that this is what the C# tag helper is doing as well the F# version wouldn't be any less good.

**Reply by dharmatech** [reply]
OK, as an experiment I've continued to roll my own input tag helper:

https://github.com/dharmatech/mvc-movie-giraffe/blob/57cc95d2d1b1acd6146b47064886dfe782456a51/TagHelpers.fs

I've converted all the instances of the input tag helper over to that and it works pretty well.

# Example

The `Movie` class of the original C# project has the following property:

```csharp
[StringLength(60, MinimumLength = 3), Required]
public string Title { get; set; }
```

Here there input tag helper is being used in the C# version of the edit view:

```cshtml
<input asp-for="Title" class="form-control" />
```

It expands into:

```html
<input 
    class="form-control"
    type="text"
    data-val="true"
    data-val-length="The field Title must be a string with a minimum length of 3 and a maximum length of 60."
    data-val-length-max="60"
    data-val-length-min="3"
    data-val-required="The Title field is required."
    id="Title"
    maxlength="60"
    name="Title"
    value="Enter the Dragon" />
```

Example of calling the F# tag helper for `input`:

```fsharp
TagHelpers.input typedefof<Movie> "Title" (string model.Title) [ _class "form-control" ]
```

That expands into:

```html
<input 
    class="form-control"
    type="text"
    data-val="true"
    data-val-length="The field Title must be a string with a minimum length of 3 and a maximum length of 60."
    data-val-length-max="60"
    data-val-length-min="3"
    maxlength="60"
    data-val-required="The Title field is required."
    id="Title"
    name="Title"
    value="Enter the Dragon">
```

**Reply by baronfel** [reply]
I think your code could get a bit more typesafe if you looked into using expressions to establish things like the title of the property, etc.

For example, here's a simple wrapper that uses quotations to extract out the property name and the property value, as well as all the custom attributes (which you already use in your existing code) for a given property of a type, but it does so in a more typed way:


```fsharp
open FSharp.Quotations
open FSharp.Quotations.Patterns

type Foo = { Name: string }

type XmlAttribute = string

// helper pattern to pull out and extract the useful info from a property getter
let (|PropInfo|_|) (e: Expr<'a>) = 
    match e with
    /// since we're only caring about properties on records, we can just pull out the relevant data
    | Patterns.PropertyGet (objInstance, propInfo, _bodyExpressions) ->
        let getter = 
            match objInstance with
            | None -> fun () -> propInfo.GetValue(null)
            | Some (ValueWithName(v, _ty, _name)) -> fun () -> propInfo.GetValue(v)
            | _ -> fun () -> box null

        Some(propInfo.Name, propInfo.PropertyType, propInfo.CustomAttributes, getter)
    | _ -> 
        None

/// have to use class members in order to use `ReflectedDefinition`, but the usage of this looks a lot
/// like a module
[<RequireQualifiedAccess>]
type Input =
    static member Of([<ReflectedDefinition>] expr: Expr<'a>, attrs: XmlAttribute list) = 
        match expr with
        | PropInfo(propName, propTy, propAttrs, getCurrentValue) ->
            let currentValue = getCurrentValue() // at this point you can inspect the property name, return type, and any custom attributes you want while also being able to get the value at any time:
            [propName; propTy.Name; string currentValue] @ attrs
        | _ -> ["got nothin'"] @ attrs


let makeForm (f: Foo) customAttrs = 
    [
        Input.Of(f.Name, customAttrs)
    ]

makeForm ({ Name = "chet" }) ["whoo"]
```

if you plop that info FSI you should see it all just work.

**Reply by dharmatech** [reply]
@baronfel,

Thanks for this example! Very interesting...

**Reply by dharmatech** [reply]
In this bit of code:

```fsharp
let (|PropInfo|_|) (e: Expr<'a>) = 
    match e with
    /// since we're only caring about properties on records, we can just pull out the relevant data
    | Patterns.PropertyGet (objInstance, propInfo, _bodyExpressions) ->
        let getter = 
            match objInstance with
            | None -> fun () -> propInfo.GetValue(null)
            | Some (ValueWithName(v, _ty, _name)) -> fun () -> propInfo.GetValue(v)
            | _ -> fun () -> box null

        Some(propInfo.Name, propInfo.PropertyType, propInfo.CustomAttributes, getter)
    | _ -> 
        None
```

I changed this line:

        Some(propInfo.Name, propInfo.PropertyType, propInfo.CustomAttributes, getter)

to:

        Some(prop_info, getter)

so as to "export" all of `prop_info` instead of just the `Name`, `PropertyType` and `CustomAttributes`.

My current experimental input tag helper is written in terms of this `PropertyInfo` so once I exposed that, I was able to drop in the body from my current implementation, rename a few things, and it worked!

# Example

Given this "model":

```fsharp
let example_movie =
    {
        Id = 123
        Title = "Abc"
        ReleaseDate = DateTime.Now
        Genre = "Xyz"
        Price = 1.23M
        Rating = "G"
    }
```

a call to the new expression-based tag helper:

```fsharp
MvcMovieGiraffe.TagHelpers.Input.Of(example_movie.Title, [])
```

results in the following:

```html
<input 
    type="text" 
    data-val="true" 
    data-val-length="The field Title must be a string with a minimum length of 3 and a maximum length of 60." 
    data-val-length-max="60" 
    data-val-length-min="3" 
    maxlength="60" 
    data-val-required="The Title field is required." 
    id="Title" 
    name="Title" 
    value="Abc">
```

Thanks so much for your example @baronfel !

**Reply by dharmatech** [reply]
# Design issue with expression-based approach

So, the approach based on @baronfel's suggestion is really cool. There is one design issue to consider however.

An example call like the one mentioned in the previous post:

```fsharp
MvcMovieGiraffe.TagHelpers.Input.Of(example_movie.Title, [])
```

of course needs a model to work (in this case `example_movie`.

If we look at the C# version of MvcMovie, the controller has the following for the `Create` method (i.e. when the user wants to create a new movie entry):

```csharp
// GET: Movies/Create
public IActionResult Create()
{
    return View();
}
```

So a model *isn't* being passed.

The create cshtml view,  has the following at the top:

```cshtml
@model MvcMovie.Models.Movie
```

but since a model isn't passed in, the `model` value is actually `null`. (I ran it in debug mode and checked the value. When you go to the 'Edit' view, `model` is not `null`.)

So the issue in mvc-movie-giraffe is that, of course, there is no model for movie in [`Views.create`](https://github.com/dharmatech/mvc-movie-giraffe/blob/main/Program.fs#L298). And so I'm not able to invoke something like this:

```fsharp
MvcMovieGiraffe.TagHelpers.Input.Of(example_movie.Title, [])
```

which requires a model. 🙂 

In mvc-movie-giraffe, since [`Movie`](https://github.com/dharmatech/mvc-movie-giraffe/blob/main/Program.fs#L34) is a record, it can't be null... So it's not clear what model to have in `Views.create`...

With the ASP.NET Core tag helpers, it seems that a null model results in the `value` attribute being set to `""`.

Any suggestions around this issue are welcome. 🙂

**Reply by baronfel** [reply]
Well yes, in that case I'd expect you to use something like `Unchecked.defaultof<Movie>.Title` in that case to get a property accessor expression that would work for you. What the aspnetcore tag helpers do is effectively that, but they rewrite the property accessors into generator functions, and then fill in the `model` at runtime and run them against a specific instance. You could do something similar as well.

Here's an example of what the 'staged' version of what I did before might look like:

```fsharp
[<RequireQualifiedAccess>]
type Input =
    static member Delayed([<ReflectedDefinition>] expr: Expr<'t -> 'a>, ?attrs: XmlAttribute list) = 
        let rec getProp e = 
            match e with
            /// since we're only caring about properties on records, we can just pull out the relevant data
            | PropertyGet (_, propInfo, _) ->
                Some(propInfo)
            | Lambda (var, expr) -> (|PropInfo|_|) expr
            | _ ->
                None
        let attrs = defaultArg attrs []
        match getProp expr with
        | Some prop ->
            let staticParts = [prop.Name; prop.PropertyType.Name] @ attrs
            fun (obj: 't) -> [prop.GetValue(obj) |> string ] @ staticParts
        | _ -> fun (obj: 't) -> ["got nothin'"] @ attrs

    
let cachedForm = Input.Delayed((fun (f: Foo) -> f.Name))

let result1 = cachedForm ({ Name = "chet" })
```

I encourage you to play around with the `Expr` modules and namespaces, there's really a lot of nice metaprogramming stuff in there!

**Reply by dharmatech** [reply]
@baronfel,

Wow... that's awesome. 🙂 

I didn't know about `Unchecked.defaultof<...>`.

The following based on your suggestion:

    MvcMovieGiraffe.TagHelpers.Input.Of(Unchecked.defaultof<Movie>.Title, [])

works fine!

Thanks again!

**Reply by dharmatech** [reply]
OK, I've updated the code to use the expression-based tag helpers.

From the edit view in the original project:

```html
<div class="form-group">
    <label asp-for="Title" class="control-label"></label>
    <input asp-for="Title" class="form-control" />
    <span asp-validation-for="Title" class="text-danger"></span>
</div>
```

The [equivalent code](https://github.com/dharmatech/mvc-movie-giraffe/blob/13848e2b79a2695ab8d74cb55c27316702ff69ef/Program.fs#L378) in the F# version:

```fsharp
div [ _class "form-group" ] [
    TagHelpers.Label.Of(model.Title, [ _class "control-label" ])

    TagHelpers.Input.Of(model.Title, [ _class "form-control" ])

    TagHelpers.SpanValidation.Of(model.Title, [ _class "text-danger" ])
]
```

Thanks to @baronfel for his help 🙂

**Reply by dharmatech** [reply]
@baronfel wrote:

> Here's an example of what the 'staged' version of what I did before might look like

For anyone who's curious about that technique, here's a short program based on @baronfel's example:

```fsharp
open FSharp.Quotations
open FSharp.Quotations.Patterns

type Foo = { Name: string }

type XmlAttribute = string

let (|PropInfo|_|) (e : Expr) =
    match e with
    | Patterns.PropertyGet (obj_instance, prop_info, _body_expressions) ->
        let getter =
            match obj_instance with
            | None                                -> fun () -> prop_info.GetValue(null)
            | Some (ValueWithName(v, _ty, _name)) -> fun () -> prop_info.GetValue(v)
            | _                                   -> fun () -> box null

        Some(prop_info)

    | _ -> None
 
[<RequireQualifiedAccess>]
type Input =
    static member Delayed([<ReflectedDefinition>] expr: Expr<'t -> 'a>, ?attrs: XmlAttribute list) = 
        let rec getProp e = 
            match e with
            | PropertyGet (_, propInfo, _) -> Some(propInfo)
            | Lambda (var, expr) ->  (|PropInfo|_|) expr
            | _ ->
                None
        let attrs = defaultArg attrs []
        match getProp expr with
        | Some prop ->
            let staticParts = [prop.Name; prop.PropertyType.Name] @ attrs
            fun (obj: 't) -> [prop.GetValue(obj) |> string ] @ staticParts
        | _ -> fun (obj: 't) -> ["no property"] @ attrs
    
let cachedForm = Input.Delayed((fun (f: Foo) -> f.Name))

let result1 = cachedForm ({ Name = "chet" })

printfn "%A" result1

```

**Reply by dharmatech** [reply]
@baronfel wrote:

> What the aspnetcore tag helpers do is effectively that, but they rewrite the property accessors into generator functions, and then fill in the `model` at runtime and run them against a specific instance. You could do something similar as well.
> 
> Here's an example of what the 'staged' version of what I did before might look like:
>     ...

So that's a pretty cool technique. As I mentioned in my last post, I did get it to run.

What I'm wondering is, how do you envision something like this being used by the tag-helpers? Here's what some calls to the tag helpers look like right now:

```fsharp
div [ _class "form-group" ] [
    TagHelpers.Label.Of(model.Title, [ _class "control-label" ])

    TagHelpers.Input.Of(model.Title, [ _class "form-control" ])

    TagHelpers.SpanValidation.Of(model.Title, [ _class "text-danger" ])
]
```

How would that change given this new technique?

**Comment by dharmatech** [comment]
The source for the tag helpers that are built-in to ASP.NET Core:

https://github.com/dotnet/aspnetcore/tree/master/src/Mvc/Mvc.TagHelpers/src

It would be nice if we could use those from F#. :-) 

Asking in aspnetcore regarding the programmatic use of the existing tag helpers:

https://github.com/dotnet/aspnetcore/discussions/28648
